jaxtyping/_decorator.py

```diff
@@ -309,6 +309,15 @@ def __init__(self, *args, **kwargs):
             # in which case make a best-effort attempt to add shape information for any
             # type errors.
 
+            # we want to detect generators, and ignore return annotations on them,
+            # to avoid issues with O(n) typechecking trying to typecheck yielded values
+            wrp = fn
+            while hasattr(wrp, "__wrapped__"):
+                wrp = wrp.__wrapped__
+                if inspect.isgeneratorfunction(wrp):
+                    fn.__annotations__["return"] = Any
```

Comment:
**Generator decorators**

Your question about a typechecking decorator using generators for some reason -- I think this could be handled by de-indenting these two lines. Do all the unwrapping, then check for being a generator.

**Comments on this approach**

That said, there are two things I don't like about this approach:
1. We're mutating an object (`fn`) that we don't own.
2. One of the intermediate decorators might have already inspected the annotations and stored a copy somewhere else. (Indeed any sane type-checking decorator will have already done this, as it will want to store a copy of any destringified annotations somewhere. That said, I don't think any of the existing typechecking decorators *actually* does this..., and I have a horrible feeling that they also trample all over the existing `__annotations__` dictionary.)

**Another possibility -- that I also don't love**

Here's one other possibility I'm contemplating, that I think should still be fairly lightweight. At this spot in the code, we have:
```python
wrp = fn
while hasattr(wrp, "__wrapped__"):
    wrp = wrp.__wrapped__
is_generator = inspect.isgeneratorfunction(wrp)
```
and then where we currently have `return fn(*args, **kwargs)`, we instead have:
```python
out = fn(*args, **kwargs):
if is_generator:
    out = _jaxtyped_generator(out)
return out
```
where we define:
```python
def _jaxtyped_generator(gen):
    while True:
        try:
            with jaxtyped("context"):
                val = next(gen)
            yield val
        except StopIteration:
            break
```
Actually, I think the above might need a little more fiddling to be sure that `.send`ing values to generators will works, but you get the idea.

**How much do we want to support?**

That said, I'm definitely not sold on my suggestion above. I don't really want to support the full gamut of generators + simple coroutines + generator courtines + `async def` coroutines + `contextlib.contextmanager`-wrapped functions + whatever else might be cooked up that I've forgotten.
([Handy reference for the first three.](https://www.integralist.co.uk/posts/python-generators/#types-of-coroutines)

Python overloads functions to mean many different things, it might be pretty difficult to support all of them.

WDYT? I'm honestly not sure of the best approach here.